src/bin/drag_and_drop_listbox_model.rs

```diff
@@ -0,0 +1,653 @@
+            }
+        }
+    }
+
+    // Public part of the RowData type. This behaves like a normal gtk-rs-style GObject
+    // binding
+    glib_wrapper! {
+        pub struct RowData(Object<subclass::simple::InstanceStruct<imp::RowData>, subclass::simple::ClassStruct<imp::RowData>, RowDataClass>);
+
+        match fn {
+            get_type => || imp::RowData::get_type().to_glib(),
+        }
+    }
+
+    // Constructor for new instances. This simply calls glib::Object::new() with
+    // initial values for our two properties and then returns the new instance
+    impl RowData {
+        pub fn new(name: &str, count: u32) -> RowData {
+            glib::Object::new(Self::static_type(), &[("name", &name), ("count", &count)])
+                .expect("Failed to create row data")
+                .downcast()
+                .expect("Created row data is of wrong type")
+        }
+    }
+}
+
+// return the previous visible row
+fn get_previous_row(list: &gtk::ListBox, row: &gtk::ListBoxRow) -> Option<gtk::ListBoxRow> {
+    if let Some(w) = list.get_row_at_index(row.get_index() - 1) {
+        if w.get_child()
+            .and_then(|w| w.downcast::<gtk::Revealer>().ok())
+            .map_or(false, |r| !r.get_reveal_child())
+        {
+            get_previous_row(list, &w)
+        } else {
+            Some(w)
+        }
+    } else {
+        None
+    }
+}
+
+// return the next visible row
+fn get_next_row(list: &gtk::ListBox, row: &gtk::ListBoxRow) -> Option<gtk::ListBoxRow> {
+    if let Some(w) = list.get_row_at_index(row.get_index() + 1) {
+        if w.get_child()
+            .and_then(|w| w.downcast::<gtk::Revealer>().ok())
+            .map_or(false, |r| !r.get_reveal_child())
+        {
+            get_next_row(list, &w)
+        } else {
+            Some(w)
+        }
+    } else {
+        None
+    }
+}
+
+fn connect_dnd_row(
+    row: &gtk::EventBox,
+    targets: &Vec<gtk::TargetEntry>,
+    channel_weak: DndChannelWeak,
+) {
+    row.drag_source_set(
+        gdk::ModifierType::BUTTON1_MASK,
+        targets,
+        gdk::DragAction::MOVE,
+    );
+
+    // Event fired when the user starts dragging, we set the ListBoxRow as the dragging icon
+    row.connect_drag_begin(move |w, ctx| {
+        if let Some(w) = w
+            .get_parent()
+            .and_then(|w| w.downcast::<gtk::Revealer>().ok())
+        {
+            let ww = w.get_allocated_width();
+            let wh = w.get_allocated_height();
+            let image = cairo::ImageSurface::create(cairo::Format::ARgb32, ww, wh).unwrap();
+            let g = cairo::Context::new(&image);
+            //TODO use theme color as the background or add a temporarily style class
+            g.set_source_rgba(1.0, 1.0, 1.0, 0.8);
+            g.rectangle(0.0, 0.0, ww as f64, wh as f64);
+            g.fill();
+
+            w.draw(&g);
+
+            //TODO fix positioning
+            //https://stackoverflow.com/questions/24844489/how-to-use-gdk-device-get-position
```

Comment:
Might be worth to ask someone who knows GTK well to go through these FIXME/TODO comments :)

Reply:
Those are there because i didn't had time to do it. But are actually not that complicated.

Reply:
Let's wait until that is done before merging then?